### ISSUE TITLE
Critical Bug: Deleting category from project renders project file and all backups inaccessible even after re-install

### ISSUE THREAD

**Issue by menelic** [ccbogel/QualCoder#265]
Deleting a category that contains codes renders this project inaccessible after closing it. Trying to reopen it leads to QualCoder freezing. Other projects can still be opened but neither the project from which the category has been deleted nor any of its backups can be accessed even after removal and reinstall of QualCoder

Steps to reproduce the behaviour:
Delete a code category that contains codes, close Qualcoder. Trying to reopen the project or any of its backups will result in QualCoder freezing.

Expected behaviour: 
It should be possible to delete a category without QualCoder being unable to access  the project and all its backups. 

This bug has been reproduced on Ubuntu 16.04 and Ubuntu 18.04

**Comment by ccbogel** [comment]
Thanks for letting me know.  I will look at this asap.

**Comment by ccbogel** [comment]
@menelic can you post the log file which might have insight into this, I am on Ubuntu 20.04 and did exactly the steps above, there was no issue that I found with QualCoder freezing or not being able to access the project.

**Comment by menelic** [comment]
thanks for the swift reply - do you mean the action log file I pasted below? This does not seem to contain much insight...note that if I open the project (eg via project > open recent project)  only the main window opens, trying to open coding > code text freezes Qualcoder and requires force quite and reopening Qualcoder to access the software. If you do not mean the action log I am posting below, please let me know how to obtain the log you need. This is a serious error that renders prior work inaccessible, so a fix is urgently needed. Maybe you could use Gnome Boxes to spin up a 18.04 within a minute (https://www.debugpoint.com/2020/05/install-use-gnome-boxes/)  to check if you can reproduce the error there? It occurred on 18.04 and was reproducible on 16.04 with survey data  imported as csv input with two fields (id and text). 

I have anonymised the output below 


This version: QualCoder 2.0
Could not detect latest release from Github
the JSON object must be str, not 'bytes'
Settings
========
Coder: default
Font: DejaVu Sans 14
Tree font size: 12
Working directory: /home/
Show IDs: False
Language: en
Timestamp format: [hh.mm.ss]
Speaker name format: []
Backup on open: True
Backup AV files: True
========
Closing project: 
========


PROJECT SUMMARY
========
Date time now: 2020-09-16 14:12
XXXX.qda
Project path: /home/XXXX.qda
Project date: 2020-09-11 12:38:47
Project memo: 
Files: 1
Cases: 6419
Code categories: 4
Codes: 11
Attributes: 0
Journals: 0
========


========
Project Opened: XXXX.qda

PROJECT SUMMARY
========
Date time now: 2020-09-16 14:12
XXXX.qda
Project path: /home/XXXXX.qda
Project date: 2020-09-11 12:38:47
Project memo: 
Files: 1
Cases: 6419
Code categories: 4
Codes: 11
Attributes: 0
Journals: 0
========

**Comment by ccbogel** [comment]
The log file is in your Home directory > .qualcoder folder > QualCoder.log

Your error was not reproducible for me because I was not using the same data as you.
Every person has a different situation and yours is different to anything I have done. You have over 6000 cases. 
So first, I assume this imported correctly and second it would have created a file from the IDs and text (go to Manage Files) and see if you can view the text file of qualitative data. 

I already suspect that the text file is really large and this is causing the code dialog to freeze.

One other option, and this depends on how you feel about this, is to privately email me either the survey to text importing and use, or a zipped project folder for me to examine.

**Comment by menelic** [comment]
I am pasting the logs below. Regarding the size of the data I think its important to know that the actual text data is quite small because these are one or two word answers - the survey had 6000 full and partial responses, Qualcoder correctly displayed only those IDs that have responses for the imported item, so the actual number of text strings is far lower than 6000. I assume the number of cases is due to the way Qualcoder imports survey csvs. Its also important to note that after import coding went smoothly, there was no lag or freezing until the code category was deleted. 

There were two logs, Qualcoder.log.1 and QualCoder.log, I am pasting both below - see esp the ingetrity error at the end of log.1

content of QualCoder.log.1 


KeyError: 'name'
2020/09/11 12:49:13 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:13 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:13 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:13 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:13 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:14 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:14 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:14 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:17 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:17 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:17 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:49:17 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/select_items.py", line 116, in data
    return QtCore.QVariant(rowitem['name'])

KeyError: 'name'
2020/09/11 12:50:04 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
2020/09/11 12:51:16 ERROR select_items.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/code_text.py", line 1528, in auto_code
    item['pos1'], item['owner'], item['memo'], item['date']))

IntegrityError: UNIQUE constraint failed: code_text.cid, code_text.fid, code_text.pos0, code_text.pos1, code_text.owner


content of QualCoder.log

the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 13:14:36 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 15:31:19 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 15:31:56 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 16:11:53 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 18:02:51 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 18:17:16 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 18:19:14 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/14 18:23:37 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/15 08:30:03 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/16 14:08:06 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/16 14:10:22 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'
the JSON object must be str, not 'bytes'
2020/09/16 14:12:59 DEBUG __main__.get_latest_github_release the JSON object must be str, not 'bytes'

**Comment by ccbogel** [comment]
@menelic I have uploaded some new files here (not in the debian). These files have changes to catch  errors associated with select_items - no 'name' key
and also  try/catch around the integrity error that was on your log file.

**Comment by ccbogel** [comment]
@menelic OK I have not used gnome-boxes yet. 
BUT: the main project opens - but the codetext does not open. Is this correct? secondly, does the Reports window open?
I now suspect the while loops that are used to fill in the codes/categories tree are running endlessly.
So I have added code to exit these after 10,000 counts. 
Can you download and try this latest update please (not in the Debian)

**Comment by menelic** [comment]
@ccbogel  thanks a lot for the fix - we actually could not reproduce the error when importing the same survey csv file, adding codes and categories and deleting categories that contain codes with the 2.1 version you provided, so on some level, the fix works. However, the project file where the error occurred (a project based on the same csv file) and all its backups created by qualcoder still cannot be opened.

Note that the version 2.1 you provided runs on 20.04 but does not start on 16.04 - there is no error reported during installation, just a notification that pip is 8. instead of 20. 

I cannot send you the original survey csv file but we have created a synthetic file with as many overall cases and the same data structure- though with more synthetic "responses" than our original file. This file takes unusually long to import - during importation qualcoder freezes - and cannot be coded because while the action log provides correct data:

Date time now:
Large N test file.qda
Project path: /home/xxx/Documents/Large N test file.qda
Project date: 
Project memo: 
Files: 0
Cases: 6419
Code categories: 0
Codes: 0
Attributes: 1
Journals: 0

the actual texts do not appear when selecting coding>code text > view file, as the "view file" window is empty. 
In addition, the synthetic file froze qualcoder and did not import when there were more synthetic responses (that is more id numbers were followed by a brief 3 word text string). This freezing did not occur with the slightly slimmer (that is more id numbers followed by empty rows)  original file where it was possible to code. I suspect that this exposes some kind of upper limit of what qualcoder can handle - and might explain the original error, because as codes were added, the file might have gotten too big to handle. When I tried to pare down the synthetic file to the size of the original file (original file ca 53 kb, synthetic file after deleting more "entries": 52 kb) this did result in correct importation of cases but there was still no text file listed in the "view file" dialogue. I am attaching the synthetic file here
[test_synthetic_survey_small.csv.zip](https://github.com/ccbogel/QualCoder/files/5264252/test_synthetic_survey_small.csv.zip)

I will get the sql statements asap - am I assuming correctly that they can just be pasted into the upper pane of the SQL statements window to get the output you are requesting?

once again, thanks for your swift reply and willingness to fix bugs, this is much appreciated!

Logs form QualCoder.og.1

2020/07/22 07:03 DEBUG __main__.read_previous_project_paths No previous projects found
2020/07/22 07:18 DEBUG __main__.read_previous_project_paths No previous projects found
2020/07/22 07:19 ERROR select_file.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/code_text.py", line 807, in view_file_dialog
    self.filename = ui.get_selected()

  File "/usr/share/qualcoder/select_file.py", line 90, in get_selected
    return self.dict_list[int(current)]

IndexError: list index out of range
2020/07/22 07:20 ERROR select_file.exception_handler Uncaught exception: Traceback (most recent call last):
  File "/usr/share/qualcoder/code_text.py", line 807, in view_file_dialog
    self.filename = ui.get_selected()

  File "/usr/share/qualcoder/select_file.py", line 90, in get_selected
    return self.dict_list[int(current)]

IndexError: list index out of range

note: this is a log from 2.0, not 2.1

**Comment by ccbogel** [comment]
Ok you need to upgrade pip in 16.04m and then perhaps re-run the pip installs for this program:
`pip install --upgrade pip`

Yes importation would take a long time to do, because of the number of rows, and for each row a case is made and then all the case text is collated into the survey text file. Yours is an unusual situation - as for qual analysis I would not have anticipated such a large survey.

It may be that the Qt framework cannot handle displaying such a large text file, but  I will use the test survey.  (on 20.04)

Re sql statement - yes that is correct in the upper pane then press the Run button.

This comment:
 "However, the project file where the error occurred (a project based on the same csv file) and all its backups created by qualcoder still cannot be opened."

(1) The project does not open at all?  or (2) does the project open but the text view does not work at all?
If (1) then it is possible the sqlite database is corrupted. You can try and open the sqlite database from the terminal: 
cd to the location of data.qda
Then run:
`sqlite3 data.qda`
Then check the tables are present:
`.tables`

Then tidy the database using:
`vacuum`

You can run some test queries e.g. (zB): 
`select * from code_text limit 30;`

You can also run the same sql statements I mentioned before.

**Comment by ccbogel** [comment]
@menelic I was able to import the test file without problems. I could view the cases and collated text file without problems.
I added codes and categories. I used the auto code feature. I delete a category. I had no problems when these actions occurred. I then closed an re opened the project.Still I had no problems occur.
![pic1](https://user-images.githubusercontent.com/5834638/93978381-436cb500-fdbf-11ea-8b90-00403b2274a8.png)
![pic4](https://user-images.githubusercontent.com/5834638/93978441-5c756600-fdbf-11ea-9879-a943f3433df8.png)
After category deleted:
![pic3](https://user-images.githubusercontent.com/5834638/93978458-60a18380-fdbf-11ea-8a78-046947a500b7.png)

One more consideration - If the text is visible but uncoded ** and you cannot see the coded text**. You might need to change the coder name to match the coder name was used to code the text. In this situation: Go to settings and select another coder name from the **Coders dropbox.**

**Comment by menelic** [comment]
@ccbogel Thank you for running such in depth tests on the file. Please note that the original error arose not from simply deleting a category but a category with codes in it. 

I have used sqlite3 to vacuum the db file in the project where such a deletion triggered the error, unfortunately that did not restore the project.

In addition, I am still unable to load the example file yousuccessfully loaded into QualCoder. It imports but when trying to code, the text does not appear on 16.04 and on 20.04. On 16.04, I get the following error when creating a code:


![screenshot qualcoder uncaught exception](https://user-images.githubusercontent.com/5934115/94009424-bd437500-fda4-11ea-8afe-42874cb0ed80.png)

The text is actually completely invisible as you can see in the screen shot. The only coder is default, so there would be no other coder to change to. But at this point, no text from the imported csv could be coded. However, I know from successfully importing the original csv that caused the error that importing and coding from csv work in my setting.

**Comment by ccbogel** [comment]
@menelic see the first picture above, the category did have codes aaa,bbb inside it. Then after the category a was deleted that is the last picture.
I an online for a little while now. So will examine this error. Did this error occur before you tried to load a text file? I see the title bar is blank.

**Comment by ccbogel** [comment]
@menelic I have added a try/catch workaround for the Uncaught Exception that is showing in you screenshot. It should also add some details to the .qualocder log file.

**Comment by menelic** [comment]
@ccbogel I have tested the latest github qualcoder master zip in 20.04 and was not able to reproduce the error - I could even open the original offending Qualcoder project in which the error occurred.. So I suggest you package 2.1 for debian release. 

One caveat: Downloading the latest github master zip installs 2.0, though I am quite sure that earlier, it did install 2.1. I cross checked this entire thread to see if you uploaded any other version anywhere - please let me know if testing with the latest Github master is sufficient. 

It definitely is different from an older 2.0 installed via deb on a 16.04 machine, because the 2.0 installed from github master zip on 20.04 could even open the original offending project. This has not been possible before, so please do release 2.1 as deb as well.

A thought on possibly saving time re: releasing for linux: Maybe a snap or a flatpack installer would enable you to automate linux releases and updates? I can't create either but i understood that to be one of their usability claims.

**Comment by ccbogel** [comment]
@menelic Thank you I am glad everything works for you now.
Yes I did have 2.1 version text in the main qualcoder.py file - I changed this back to 2.0 as I realised it was not appropriate until a 2.1 version was released.
Survey Import:  have added a tooltip for changing column names and for suggestion of changing from Character to Qualitative. I have added some text after 'Changes...' as you recommended.
I had a look at snap and flatpack a long time ago. I found it too hard to follow the instructions to create them.
I will do another deb after a couple of other issues are resolved.
I have noticed QualCoder has been gaining more interest via the Wordpress site and the alternativeto website.
